connect-inject/endpoints_controller.go

```diff
@@ -710,19 +716,94 @@ func (r *EndpointsController) deregisterServiceOnAllAgents(ctx context.Context,
 						r.Log.Error(err, "failed to deregister service instance", "id", svcID)
 						return err
 					}
+
+					if r.AuthMethod != "" {
+						r.Log.Info("deleting ACL token for service instance", "id", svcID)
+						// We don't check that these keys exist in the map because we expect that they always do by the time we're
+						// looking at this service registration.
```

Comment:
We definitely expect the MetaKeyKubeNS to exist because we get services using the namespace in line 701, but if MetaKeyPodName is missing (even just due to a user error) do we want this code to panic, or just log the error in deleting the ACL token? 

Maybe since Meta keys are something a user could potentially have automation around, if they wrote something with a bug that removed MetaKeyPodName I'm sure it would cause lots of problems, but maybe just worth thinking if we want a panic in that case or a handled error. I think I'm just leaning slightly towards handled error but if its really not possible to get to this line in that case, its totally fine if you think otherwise!

Reply:
That's a great question!

I've changed `serviceRegistration.Meta[MetaKeyKubeNS]` to `k8sSvcNamespace` which is already passed as an argument to that function. Thank you for pointing that out!

Thinking about the `MetaKeyPodName` behavior, I'm leaning towards keeping it as is because:
1. Golang maps won't panic if the key doesn't exist: https://play.golang.org/p/yJe52t5c0W6 . Instead, they return an empty string.
2. If pod name is an empty string, then the `deleteACLTokenForService` won't be able to find the token for the service because there will be no token with metadata `{"pod": "/<the value of k8sSvcNamespace>"` and will result in a no-op. I think this is the correct behavior since we can't determine the token that needs to be deleted. We also shouldn't error because that will trigger reconciliation again.

One thing that could be useful though is logging that the pod name meta key is missing if it's not found. WDYT?

Reply:
Actually, your question made me realize that I wasn't handling a case where you have orphaned tokens, i.e. tokens for which there are no service registrations. As a result of fixing that, I no longer need to access pod name from the registration and instead check whether the pod exists in kubernetes. That seems like a better solution to me and it also made the code a bit easier!

Thank you so much for asking such great questions ❤️